shared/src/vm/mod.rs

```diff
@@ -25,7 +25,7 @@ const UNTRUSTED_WASM_FEATURES: WasmFeatures = WasmFeatures {
     relaxed_simd: false,
     threads: false,
     tail_call: false,
-    floats: false,
+    floats: true,
```

Comment:
I don't think we want to allow floats in wasm as it's allowed to give non deterministic results for NaNs. It's not needed for the gas, is it?

Reply:
Ah yes, sorry I should have left a note on this one. So basically I reworked the way we call `fetch_or_compile` and `compile_and_fetch` and also the implementation of `fetch_or_compile`. This is because `compile_or_fetch` is eventually called in any case (either the code being the actual code or only the hash of it). So I basically rewrote it to call `fetch_or_compile` in any case (this ease the gas accounting operation that must be done in any case and can therefore be done in this function only) and then let this function call `compile_or_fetch` in turn.

Now, in the process  of doing this, I also moved the call to `validate_untrusted_wasm` because, at the moment, we are validating the code only when we receive an hash but I think we should do the opposite. When we get an hash, if the precompiled module is not in cache, we end up reading the code from storage. But I think there should be no need to validate code placed in storage (validation should happen when we write the code to storage which I think we are not doing at the moment). If instead of an hash we get some code, we need to validate it before running it (because if the tx is not whitelisted it would only be discarded by the vps, so after it's run).

Now, to the point of the floats. In the compilation of the ibc transaction we emit some `f64` operations (I checked it with `wasm2wat` on the code emitted using both `build-wasm-scripts` and `build-wasm-scripts-docker` from different releases, even in the current version of main these instructions are there). The rearrangement of the compilation steps I mentioned before brought up this thing which, usually, doesn't manifest itself because the precompiled wasm module happens to always be in cache and so skips the validation part. To double check this, in the main branch, I removed the load from cache and forced the compilation of the module every time and indeed, the validation step rejected the ibc transaction because of forbidden floating point operations.

I had a look at the ibc code and it looks that these f64 operations where coming from one of ibc dependencies, in particular from a single insert in an HashMap which, for some strange reasons, triggered the generation of f64 operations in wasm, so I simply allowed floats to speed up the development process since I think I cannot solve this issue by myself

Reply:
Ah, I see. The change makes sense, thanks! 

I had a look at tx_ibc (I left the symbols in the wasm [by removing `RUSTFLAGS='-C link-arg=-s'` from makefile]) and found that the `f64` usage comes from `serde_json` and it's compiled into a number parses even if floats are not used :/  It's being called from e.g. `send_transfer_execute` imported from ibc-rs. If we want to remove the usage of floats we have to use a fork (https://github.com/serde-rs/json/issues/567). There's one from Near https://github.com/nearprotocol/json, it's a bit behind so might need to rebase on later version.

cc @yito88

Reply:
https://github.com/cosmos/ibc-rs/issues/850